### PR TITLE
[crypto] remove compat and update Uniform

### PR DESCRIPTION
**Pull request by davidiw**
compat was the last part of the legacy API, this removes it from both ed25519 and x25519
I also tweaked Uniform to rename generate_for_testing to generate and introduced a generate_for_testing for convenience (it uses the TEST_SEED). I also added a very clear note on the lack of safety involving generate, but it may very well be used in a couple production cases, and we cannot leave it named as is...
moved proptest code out as it was a clean replacement for the compat interfaces

### CODE REVIEW

File: admission_control/admission-control-service/src/unit_tests/admission_control_service_test.rs [hunk 1]
```diff
@@ -34,16 +34,15 @@ fn submit_transaction(
     ac_service: &MockAdmissionControlService,
     are_keys_valid: bool,
 ) -> SubmitTransactionResponse {
-    let keypair = compat::generate_keypair(None);
+    let mut rng = ::rand::rngs::StdRng::from_seed(TEST_SEED);
```
Comment (kchalkias) [inline admission_control/admission-control-service/src/unit_tests/admission_control_service_test.rs]: This is `generate_for_testing`

File: client/libra-dev/src/transaction.rs [hunk 2]
```diff
@@ -390,283 +394,278 @@ pub unsafe extern "C" fn libra_LibraSignedTransaction_from(
...
-            expiration_time_secs,
-            buf_ptr,
-            &mut len,
-        )
-    };
+    /// Generate a Signed Transaction and deserialize
+    #[test]
+    fn test_lcs_signed_transaction() {
+        // generate key pair
+        let mut rng = StdRng::from_seed(TEST_SEED);
```
Comment (kchalkias) [inline client/libra-dev/src/transaction.rs]: ditto `generate_for_testing`

File: client/libra-dev/src/transaction.rs [hunk 3]
```diff
@@ -390,283 +394,278 @@ pub unsafe extern "C" fn libra_LibraSignedTransaction_from(
...
-fn test_libra_raw_transaction_bytes_from() {
-    use lcs::from_bytes;
-    use libra_crypto::{hash::CryptoHash, test_utils::TEST_SEED, traits::SigningKey};
-    use libra_types::transaction::{SignedTransaction, TransactionArgument};
-    use rand::{rngs::StdRng, SeedableRng};
+    /// Generate a Raw Transaction, sign it, then deserialize
+    #[test]
+    fn test_libra_raw_transaction_bytes_from() {
+        // generate key pair
+        let mut rng = StdRng::from_seed(TEST_SEED);
```
Comment (kchalkias) [inline client/libra-dev/src/transaction.rs]: ditto `generate_for_testing`

File: config/src/keys.rs [hunk 4]
```diff
@@ -82,7 +82,7 @@ where
 {
     fn default() -> Self {
         let mut rng = StdRng::from_seed(TEST_SEED);
-        let private_key = T::generate_for_testing(&mut rng);
+        let private_key = T::generate(&mut rng);
```
Comment (kchalkias) [inline config/src/keys.rs]: `generate_for_testing`

File: crypto/crypto/src/unit_tests/multi_ed25519_test.rs [hunk 5]
```diff
@@ -18,9 +18,11 @@ use rand::{rngs::StdRng, SeedableRng};
 static MESSAGE_HASH: Lazy<HashValue> = Lazy::new(|| HashValue::from_sha3_256(b"Test Message"));
 
 // Helper function to generate N key pairs.
-fn generate_key_pairs(n: usize) -> Vec<(Ed25519PrivateKey, Ed25519PublicKey)> {
+fn generate_keys(n: usize) -> Vec<Ed25519PrivateKey> {
     let mut rng = StdRng::from_seed(TEST_SEED);
-    (0..n).map(|_| compat::generate_keypair(&mut rng)).collect()
+    (0..n)
+        .map(|_| Ed25519PrivateKey::generate(&mut rng))
```
Comment (kchalkias) [inline crypto/crypto/src/unit_tests/multi_ed25519_test.rs]: +1, we shouldn't use `generate_for_testing` here as many keys are derived from the same "starting" seed, but the seed should be different between them.

File: execution/executor/tests/storage_integration_test.rs [hunk 6]
```diff
@@ -191,7 +195,7 @@ fn test_reconfiguration() {
 
     // rotate the validator's connsensus pubkey to trigger a reconfiguration
     let mut rng = ::rand::rngs::StdRng::from_seed(TEST_SEED);
-    let (_, new_pubkey) = compat::generate_keypair(&mut rng);
+    let new_pubkey = Ed25519PrivateKey::generate(&mut rng).public_key();
```
Comment (kchalkias) [inline execution/executor/tests/storage_integration_test.rs]: Good point, we cannot use `generate_for_testing` here, as we verify (assert) the seed.

File: network/src/connectivity_manager/test.rs [hunk 7]
```diff
@@ -72,8 +75,8 @@ fn setup_conn_mgr(
 fn gen_peer() -> (PeerId, NetworkPublicKeys) {
     let peer_id = PeerId::random();
     let mut rng = StdRng::from_seed(TEST_SEED);
-    let (_, signing_public_key) = compat::generate_keypair(&mut rng);
-    let (_, identity_public_key) = x25519::compat::generate_keypair(&mut rng);
+    let signing_public_key = Ed25519PrivateKey::generate(&mut rng).public_key();
```
Comment (kchalkias) [inline network/src/connectivity_manager/test.rs]: +1, we can't use `generate_for_testing` here as we generate more than one keys.

File: network/src/protocols/discovery/test.rs [hunk 8]
```diff
@@ -108,8 +111,8 @@ async fn expect_address_update(
 
 fn generate_network_pub_keys_and_signer() -> (NetworkPublicKeys, Ed25519PrivateKey) {
     let mut rng = StdRng::from_seed(TEST_SEED);
-    let signing_priv_key = Ed25519PrivateKey::generate_for_testing(&mut rng);
-    let (_, identity_pub_key) = x25519::compat::generate_keypair(&mut rng);
+    let signing_priv_key = Ed25519PrivateKey::generate(&mut rng);
```
Comment (kchalkias) [inline network/src/protocols/discovery/test.rs]: `generate_for_testing`

File: network/src/protocols/network/dummy.rs [hunk 9]
```diff
@@ -102,16 +105,16 @@ pub fn setup_network() -> DummyNetwork {
 
     // Setup keys for dialer.
     let mut rng = StdRng::from_seed(TEST_SEED);
-    let (dialer_signing_private_key, dialer_signing_public_key) =
-        compat::generate_keypair(&mut rng);
-    let (dialer_identity_private_key, dialer_identity_public_key) =
-        x25519::compat::generate_keypair(&mut rng);
+    let dialer_signing_private_key = Ed25519PrivateKey::generate(&mut rng);
```
Comment (kchalkias) [inline network/src/protocols/network/dummy.rs]: +1 we cannot use `generate_for_testing` here.

File: types/src/validator_signer.rs [hunk 10]
```diff
@@ -69,7 +69,7 @@ impl ValidatorSigner {
         let mut address = [0; AccountAddress::LENGTH];
         address[0] = num;
         let mut rng = StdRng::from_seed(TEST_SEED);
-        let private_key = Ed25519PrivateKey::generate_for_testing(&mut rng);
+        let private_key = Ed25519PrivateKey::generate(&mut rng);
```
Comment (kchalkias) [inline types/src/validator_signer.rs]: `generate_for_testing`

File: types/src/validator_signer.rs [hunk 11]
```diff
@@ -85,9 +85,7 @@ pub mod proptests {
         prop_oneof![
             // The no_shrink here reflects that particular keypair choices out
             // of random options are irrelevant.
-            LazyJust::new(
-                || Ed25519PrivateKey::generate_for_testing(&mut StdRng::from_seed(TEST_SEED))
-            ),
+            LazyJust::new(|| Ed25519PrivateKey::generate(&mut StdRng::from_seed(TEST_SEED))),
```
Comment (kchalkias) [inline types/src/validator_signer.rs]: `generate_for_testing`

File: secure/storage/src/value.rs [hunk 12]
```diff
@@ -59,7 +59,7 @@ mod tests {
     #[test]
     fn ed25519_private_key() {
         let mut rng = StdRng::from_seed([13u8; 32]);
-        let value = Ed25519PrivateKey::generate_for_testing(&mut rng);
+        let value = Ed25519PrivateKey::generate(&mut rng);
```
Comment (kchalkias) [inline secure/storage/src/value.rs]: In theory we could use `generate_for_testing`, no need for another custom test rng

File: secure/storage/src/tests/suite.rs [hunk 13]
```diff
@@ -383,7 +383,7 @@ fn test_create_sign_rotate_sign(storage: &mut dyn Storage) {
 
 fn create_ed25519_key_for_testing() -> Ed25519PrivateKey {
     let mut rng = StdRng::from_seed([13u8; 32]);
-    Ed25519PrivateKey::generate_for_testing(&mut rng)
+    Ed25519PrivateKey::generate(&mut rng)
```
Comment (kchalkias) [inline secure/storage/src/tests/suite.rs]: Why we didn't use`generate_for_testing` here? Is there a specific reason to have this extra test rng?

File: admission_control/admission-control-service/src/unit_tests/admission_control_service_test.rs (hunk 1 above)
Comment (davidiw) [reply admission_control/admission-control-service/src/unit_tests/admission_control_service_test.rs]: They generate secondary key if "are_keys_valid" is false. Actually if you look the test code was actually buggy so I wonder if it even tested anything... probably worth flagging to these folks, but I think this code has a limited shelf-life.

File: admission_control/admission-control-service/src/unit_tests/admission_control_service_test.rs [hunk 14]
```diff
@@ -34,16 +34,15 @@ fn submit_transaction(
...
+    let private_key = Ed25519PrivateKey::generate(&mut rng);
     let public_key = if are_keys_valid {
-        keypair.1
+        private_key.public_key()
     } else {
-        let mut rng = ::rand::rngs::StdRng::from_seed(TEST_SEED);
-        let test_key = compat::generate_keypair(&mut rng);
-        test_key.1
+        Ed25519PrivateKey::generate(&mut rng).public_key()
```
Comment (davidiw) [inline admission_control/admission-control-service/src/unit_tests/admission_control_service_test.rs]: see, the rng is used again to generate a second, different key.

File: crypto/crypto/src/unit_tests/multi_ed25519_test.rs (hunk 5 above)
Comment (davidiw) [reply crypto/crypto/src/unit_tests/multi_ed25519_test.rs]: note this is similar to the AC above.

File: execution/executor/tests/storage_integration_test.rs (hunk 6 above)
Comment (davidiw) [reply execution/executor/tests/storage_integration_test.rs]: actually, this one can be compressed, the test is in another test!

File: network/src/protocols/discovery/test.rs (hunk 8 above)
Comment (davidiw) [reply network/src/protocols/discovery/test.rs]: This one is a little nuanced but the test shouldn't depend on the uniqueness of keys.. but we do generate secondary EC key afterward.

File: secure/storage/src/tests/suite.rs (hunk 13 above)
Comment (davidiw) [reply secure/storage/src/tests/suite.rs]: good question, I should have cleaned this up... historically there was no "modern" interface for generating a Ed25519PrivateKey hence you see this fragmentation.

File: types/src/validator_signer.rs [hunk 15]
```diff
@@ -59,7 +59,7 @@ impl ValidatorSigner {
         let mut rng = StdRng::from_seed(opt_rng_seed.into().unwrap_or(TEST_SEED));
         Self::new(
             AccountAddress::random(),
-            Ed25519PrivateKey::generate_for_testing(&mut rng),
+            Ed25519PrivateKey::generate(&mut rng),
```
Comment (davidiw) [inline types/src/validator_signer.rs]: I wonder about this one....

Comment (davidiw) [reply types/src/validator_signer.rs]: Nope, it is used elsewhere :/

File: testsuite/tests/libratest/smoke_test.rs [hunk 16]
```diff
@@ -642,9 +647,8 @@ fn test_external_transaction_signer() {
     // generate key pair
     let mut seed: [u8; 32] = [0u8; 32];
     seed[..4].copy_from_slice(&[1, 2, 3, 4]);
-    let key_pair = compat::generate_keypair(None);
-    let private_key = key_pair.0;
-    let public_key = key_pair.1;
+    let private_key = Ed25519PrivateKey::generate_for_testing();
```
Comment (kchalkias) [inline testsuite/tests/libratest/smoke_test.rs]: `seed` is not used.

**Pull request closed without merging**